### PR TITLE
Replacing strftime with internal implementation

### PR DESCRIPTION
**Pull request by phprus**
Proposal: https://github.com/fmtlib/fmt/pull/2500#issuecomment-920078450
Discussion: Issue #2541 

All locale independent formats are implemented native with significant performance increase (С++11: 2-4 times, C++17: up to 13 times).
For locale dependent formats, strftime is used.

Draft because:

- [x] Need check ISO week/year algorithm.

#### Performance regression on Windows, because ``ucrt`` ``strftime`` is slow by design.

Possible solution for Windows regression: split processing into 2 way: fast path and slow path.
Slow path: unrecognized format string (non-standard extensions) or locale-dependent (except  "C" locale?) formats in the format string.
Else: Fast path.
Pass the slow path formatting string to ``strftime`` or ``std::time_put``. Fast path - internal implementation.

### CODE REVIEW

**Comment by toughengineer** [comment]
What is the goal of these changes?

**Comment by toughengineer** [comment]
So here are the results of quick and dirty tests of current trunk and you implementation at fd23701 (timings are in nanoseconds):
<table>
<thead>
  <tr>
    <th></th>
    <th>%Y</th>
    <th>%F %T</th>
    <th>%Y-%m-%d %H:%M:%S</th>
    <th>%a %b %d %T %Y</th>
    <th>%a, %d %b %Y %T %z</th>
  </tr>
</thead>
<tbody>
  <tr>
    <td>baseline</td>
    <td>290</td>
    <td>436</td>
    <td>452</td>
    <td>570</td>
    <td>534</td>
  </tr>
  <tr>
    <td>baseline_compile</td>
    <td>265</td>
    <td>388</td>
    <td>414</td>
    <td>532</td>
    <td>478</td>
  </tr>
  <tr>
    <td>fd23701</td>
    <td>40.5</td>
    <td>66</td>
    <td>119</td>
    <td>647</td>
    <td>961</td>
  </tr>
  <tr>
    <td>fd23701_compile</td>
    <td>16.3</td>
    <td>29.6</td>
    <td>50.5</td>
    <td>544</td>
    <td>904</td>
  </tr>
</tbody>
</table>

The penultimate format string is a shorter form from this [stackoverflow answer](https://stackoverflow.com/a/10289094).\
The last one is an RFC 2822 example from [`strftime` man page](https://man7.org/linux/man-pages/man3/strftime.3.html#EXAMPLES).\
I.e. these are _not uncommon_ options.

Now for simple cases your implementation saves from 250 to 370 ns.\
But for the more complex cases your implementation is slower from 70 to 420 ns (-ish).\
This means that if a user happens to use one of the simple options, he is lucky and it runs faster. That's great!\
But if the user is unlucky and uses more complex options, your implementation penalizes the user performance-wise. It runs **almost twice as slow** for the RFC 2822 example compared to the current trunk.

Basically you made fast cases faster, but slow cases much slower.

That's why I asked about the goal of these changes.\
If it is performance improvements, which in some cases there are, what is the justification to favor simple, fast cases, and slow down complex, _slower_ cases?

**Comment by toughengineer** [comment]
Windows 10\
MSVC 16.11.4

```
Run on (8 X 3600 MHz CPU s)
CPU Caches:
  L1 Data 32 KiB (x4)
  L1 Instruction 32 KiB (x4)
  L2 Unified 256 KiB (x4)
  L3 Unified 8192 KiB (x1)
```
<details>
<summary>copy paste of all runs</summary>

C++11
```
D:\dev\fmt-bench\out\build\x64-Release>fmt_test_old.exe
2021-10-13T22:47:16+03:00
Running fmt_test_old.exe
Run on (8 X 3600 MHz CPU s)
CPU Caches:
  L1 Data 32 KiB (x4)
  L1 Instruction 32 KiB (x4)
  L2 Unified 256 KiB (x4)
  L3 Unified 8192 KiB (x1)
------------------------------------------------------------------------
Benchmark                              Time             CPU   Iterations
------------------------------------------------------------------------
FMTFormatter_Y                       284 ns          285 ns      2357895
FMTFormatterCompile_Y                287 ns          289 ns      2488889
FMTFormatter_full                    431 ns          430 ns      1600000
FMTFormatterCompile_full             442 ns          449 ns      1600000
FMTFormatter_full2                   415 ns          419 ns      1792000
FMTFormatterCompile_full2            403 ns          399 ns      1723077
FMTFormatter_complex_1               446 ns          449 ns      1600000
FMTFormatterCompile_complex_1        447 ns          449 ns      1600000
FMTFormatter_complex_2               527 ns          530 ns      1120000
FMTFormatterCompile_complex_2        533 ns          547 ns      1000000

D:\dev\fmt-bench\out\build\x64-Release>fmt_test_new.exe
2021-10-13T22:47:32+03:00
Running fmt_test_new.exe
Run on (8 X 3600 MHz CPU s)
CPU Caches:
  L1 Data 32 KiB (x4)
  L1 Instruction 32 KiB (x4)
  L2 Unified 256 KiB (x4)
  L3 Unified 8192 KiB (x1)
------------------------------------------------------------------------
Benchmark                              Time             CPU   Iterations
------------------------------------------------------------------------
FMTFormatter_Y                      57.4 ns         57.2 ns     11200000
FMTFormatterCompile_Y               57.0 ns         57.8 ns     10000000
FMTFormatter_full                    138 ns          136 ns      4480000
FMTFormatterCompile_full             139 ns          138 ns      4977778
FMTFormatter_full2                  87.6 ns         87.9 ns      7466667
FMTFormatterCompile_full2           87.7 ns         87.9 ns      7466667
FMTFormatter_complex_1               607 ns          614 ns      1120000
FMTFormatterCompile_complex_1        623 ns          628 ns      1120000
FMTFormatter_complex_2               929 ns          921 ns       746667
FMTFormatterCompile_complex_2        939 ns          942 ns       746667
```
C++17
```
D:\dev\fmt-bench\out\build\x64-Release>fmt_test_old.exe
2021-10-13T22:51:33+03:00
Running fmt_test_old.exe
Run on (8 X 3600 MHz CPU s)
CPU Caches:
  L1 Data 32 KiB (x4)
  L1 Instruction 32 KiB (x4)
  L2 Unified 256 KiB (x4)
  L3 Unified 8192 KiB (x1)
------------------------------------------------------------------------
Benchmark                              Time             CPU   Iterations
------------------------------------------------------------------------
FMTFormatter_Y                       291 ns          289 ns      2488889
FMTFormatterCompile_Y                248 ns          246 ns      2800000
FMTFormatter_full                    436 ns          433 ns      1659259
FMTFormatterCompile_full             368 ns          368 ns      1866667
FMTFormatter_full2                   406 ns          408 ns      1723077
FMTFormatterCompile_full2            351 ns          352 ns      1866667
FMTFormatter_complex_1               449 ns          445 ns      1544828
FMTFormatterCompile_complex_1        384 ns          385 ns      1866667
FMTFormatter_complex_2               527 ns          531 ns      1000000
FMTFormatterCompile_complex_2        456 ns          455 ns      1544828

D:\dev\fmt-bench\out\build\x64-Release>fmt_test_new.exe
2021-10-13T22:51:47+03:00
Running fmt_test_new.exe
Run on (8 X 3600 MHz CPU s)
CPU Caches:
  L1 Data 32 KiB (x4)
  L1 Instruction 32 KiB (x4)
  L2 Unified 256 KiB (x4)
  L3 Unified 8192 KiB (x1)
------------------------------------------------------------------------
Benchmark                              Time             CPU   Iterations
------------------------------------------------------------------------
FMTFormatter_Y                      54.3 ns         54.7 ns     10000000
FMTFormatterCompile_Y               21.3 ns         21.3 ns     34461538
FMTFormatter_full                    152 ns          151 ns      4977778
FMTFormatterCompile_full            45.7 ns         45.5 ns     15448276
FMTFormatter_full2                  92.6 ns         92.1 ns      7466667
FMTFormatterCompile_full2           32.4 ns         32.8 ns     22400000
FMTFormatter_complex_1               621 ns          628 ns      1120000
FMTFormatterCompile_complex_1        496 ns          500 ns      1000000
FMTFormatter_complex_2               933 ns          942 ns       746667
FMTFormatterCompile_complex_2        798 ns          802 ns       896000
```
C++20
```
D:\dev\fmt-bench\out\build\x64-Release>fmt_test_old.exe
2021-10-13T22:55:29+03:00
Running fmt_test_old.exe
Run on (8 X 3600 MHz CPU s)
CPU Caches:
  L1 Data 32 KiB (x4)
  L1 Instruction 32 KiB (x4)
  L2 Unified 256 KiB (x4)
  L3 Unified 8192 KiB (x1)
------------------------------------------------------------------------
Benchmark                              Time             CPU   Iterations
------------------------------------------------------------------------
FMTFormatter_Y                       290 ns          292 ns      2357895
FMTFormatterCompile_Y                250 ns          251 ns      2800000
FMTFormatter_full                    443 ns          449 ns      1600000
FMTFormatterCompile_full             370 ns          369 ns      1947826
FMTFormatter_full2                   411 ns          414 ns      1659259
FMTFormatterCompile_full2            345 ns          345 ns      1947826
FMTFormatter_complex_1               456 ns          446 ns      1120000
FMTFormatterCompile_complex_1        381 ns          385 ns      1866667
FMTFormatter_complex_2               551 ns          544 ns      1120000
FMTFormatterCompile_complex_2        453 ns          449 ns      1600000

D:\dev\fmt-bench\out\build\x64-Release>fmt_test_new.exe
2021-10-13T22:55:42+03:00
Running fmt_test_new.exe
Run on (8 X 3600 MHz CPU s)
CPU Caches:
  L1 Data 32 KiB (x4)
  L1 Instruction 32 KiB (x4)
  L2 Unified 256 KiB (x4)
  L3 Unified 8192 KiB (x1)
------------------------------------------------------------------------
Benchmark                              Time             CPU   Iterations
------------------------------------------------------------------------
FMTFormatter_Y                      53.0 ns         53.0 ns     11200000
FMTFormatterCompile_Y               21.4 ns         21.3 ns     34461538
FMTFormatter_full                    145 ns          145 ns      5600000
FMTFormatterCompile_full            49.9 ns         50.8 ns     14451613
FMTFormatter_full2                  91.3 ns         92.8 ns      6400000
FMTFormatterCompile_full2           32.5 ns         32.2 ns     21333333
FMTFormatter_complex_1               631 ns          628 ns      1120000
FMTFormatterCompile_complex_1        501 ns          500 ns      1000000
FMTFormatter_complex_2               947 ns          942 ns       746667
FMTFormatterCompile_complex_2        828 ns          820 ns       896000
```

</details>

<table>
<thead>
  <tr>
    <th rowspan="2"></th>
    <th colspan="2">C++11</th>
    <th colspan="2">C++17</th>
    <th colspan="2">C++20</th>
  </tr>
  <tr>
    <th>old</th>
    <th>new</th>
    <th>old</th>
    <th>new</th>
    <th>old</th>
    <th>new</th>
  </tr>
</thead>
<tbody>
  <tr>
    <td>FMTFormatter_Y</td>
    <td>284</td>
    <td>57.4</td>
    <td>291</td>
    <td>54.3</td>
    <td>290</td>
    <td>53</td>
  </tr>
  <tr>
    <td>FMTFormatterCompile_Y</td>
    <td>287</td>
    <td>57</td>
    <td>248</td>
    <td>21.3</td>
    <td>250</td>
    <td>21.4</td>
  </tr>
  <tr>
    <td>FMTFormatter_full</td>
    <td>431</td>
    <td>138</td>
    <td>436</td>
    <td>152</td>
    <td>443</td>
    <td>145</td>
  </tr>
  <tr>
    <td>FMTFormatterCompile_full</td>
    <td>442</td>
    <td>139</td>
    <td>368</td>
    <td>45.7</td>
    <td>370</td>
    <td>49.9</td>
  </tr>
  <tr>
    <td>FMTFormatter_full2</td>
    <td>415</td>
    <td>87.6</td>
    <td>406</td>
    <td>92.6</td>
    <td>411</td>
    <td>91.3</td>
  </tr>
  <tr>
    <td>FMTFormatterCompile_full2</td>
    <td>403</td>
    <td>87.7</td>
    <td>351</td>
    <td>32.4</td>
    <td>345</td>
    <td>32.5</td>
  </tr>
  <tr>
    <td>FMTFormatter_complex_1</td>
    <td>446</td>
    <td>607</td>
    <td>449</td>
    <td>621</td>
    <td>456</td>
    <td>631</td>
  </tr>
  <tr>
    <td>FMTFormatterCompile_complex_1</td>
    <td>447</td>
    <td>623</td>
    <td>384</td>
    <td>496</td>
    <td>381</td>
    <td>501</td>
  </tr>
  <tr>
    <td>FMTFormatter_complex_2</td>
    <td>527</td>
    <td>929</td>
    <td>527</td>
    <td>933</td>
    <td>551</td>
    <td>947</td>
  </tr>
  <tr>
    <td>FMTFormatterCompile_complex_2</td>
    <td>533</td>
    <td>939</td>
    <td>456</td>
    <td>798</td>
    <td>453</td>
    <td>828</td>
  </tr>
</tbody>
</table>

**Comment by phprus** [comment]
Thanks!
What are your locale settings?

**Comment by phprus** [comment]
Windows ucrt ``strftime`` is very slow by design, because has memory allocation, convert format string from ``char`` to ``wchar_t`` and convert result string from ``wchar_t`` to ``char``.

Replacing ``strftime`` with ``std::time_put<wchar_t, ...>`` with convert the result to ``char`` can help speed up this code on Windows. And for the ``C`` locale, may not call ``std::time_put``, but use the built-in arrays of month names, etc.

**Comment by vitaut** [comment]
I wouldn't worry too much about Windows, we can optimize exotic formats there later.

**Comment by phprus** [comment]
Done

**Comment by vitaut** [comment]
Thank you!

**Comment by vitaut** [comment]
Looks like there is a UB in `iso_year_weeks`. Here's a stack trace from oss-fuzz:

```
/src/fmt/include/fmt/chrono.h:1428:20: runtime error: signed integer overflow: -1996866182 + -499216545 cannot be represented in type 'int'
--
  | #0 0x4d58c5 in fmt::v8::detail::tm_writer<fmt::v8::appender, char>::iso_year_weeks(int) const fmt/include/fmt/chrono.h:1428:20
  | #1 0x4d5a0e in fmt::v8::detail::tm_writer<fmt::v8::appender, char>::tm_iso_week_of_year() const fmt/include/fmt/chrono.h:1450:13
  | #2 0x4d2c99 in fmt::v8::detail::tm_writer<fmt::v8::appender, char>::on_iso_week_of_year(fmt::v8::detail::numeric_system) fmt/include/fmt/chrono.h:1622:26
  | #3 0x4d2c99 in char const* fmt::v8::detail::parse_chrono_format<char, fmt::v8::detail::tm_writer<fmt::v8::appender, char>&>(char const*, char const*, fmt::v8::detail::tm_writer<fmt::v8::appender, char>&) fmt/include/fmt/chrono.h:620:15
  | #4 0x4d1a34 in decltype(fp0.out()) fmt::v8::formatter<tm, char, void>::format<fmt::v8::basic_format_context<fmt::v8::appender, char> >(tm const&, fmt::v8::basic_format_context<fmt::v8::appender, char>&) const fmt/include/fmt/chrono.h:1768:7
  | #5 0x4d196f in void fmt::v8::detail::value<fmt::v8::basic_format_context<fmt::v8::appender, char> >::format_custom_arg<tm, fmt::v8::formatter<tm, char, void> >(void*, fmt::v8::basic_format_parse_context<char, fmt::v8::detail::error_handler>&, fmt::v8::basic_format_context<fmt::v8::appender, char>&) fmt/include/fmt/core.h:1241:22
  | #6 0x4c6c74 in fmt::v8::basic_format_arg<fmt::v8::basic_format_context<fmt::v8::appender, char> >::handle::format(fmt::v8::basic_format_parse_context<char, fmt::v8::detail::error_handler>&, fmt::v8::basic_format_context<fmt::v8::appender, char>&) const fmt/include/fmt/core.h:1503:7
  | #7 0x4c77c2 in fmt::v8::detail::custom_formatter<char>::operator()(fmt::v8::basic_format_arg<fmt::v8::basic_format_context<fmt::v8::appender, char> >::handle) const fmt/include/fmt/format.h:2197:7
  | #8 0x4c77c2 in decltype(fp(0)) fmt::v8::visit_format_arg<fmt::v8::detail::custom_formatter<char>, fmt::v8::basic_format_context<fmt::v8::appender, char> >(fmt::v8::detail::custom_formatter<char>&&, fmt::v8::basic_format_arg<fmt::v8::basic_format_context<fmt::v8::appender, char> > const&) fmt/include/fmt/core.h:1567:12
  | #9 0x4c77c2 in void fmt::v8::detail::vformat_to<char>(fmt::v8::detail::buffer<char>&, fmt::v8::basic_string_view<char>, fmt::v8::basic_format_args<fmt::v8::basic_format_context<std::__1::conditional<std::is_same<fmt::v8::type_identity<char>::type, char>::value, fmt::v8::appender, std::__1::back_insert_iterator<fmt::v8::detail::buffer<fmt::v8::type_identity<char>::type> > >::type, fmt::v8::type_identity<char>::type> >, fmt::v8::detail::locale_ref)::format_handler::on_format_specs(int, char const*, char const*) fmt/include/fmt/format.h:2938:9
  | #10 0x4c6f3c in char const* fmt::v8::detail::parse_replacement_field<char, void fmt::v8::detail::vformat_to<char>(fmt::v8::detail::buffer<char>&, fmt::v8::basic_string_view<char>, fmt::v8::basic_format_args<fmt::v8::basic_format_context<std::__1::conditional<std::is_same<fmt::v8::type_identity<char>::type, char>::value, fmt::v8::appender, std::__1::back_insert_iterator<fmt::v8::detail::buffer<fmt::v8::type_identity<char>::type> > >::type, fmt::v8::type_identity<char>::type> >, fmt::v8::detail::locale_ref)::format_handler&>(char const*, char const*, void fmt::v8::detail::vformat_to<char>(fmt::v8::detail::buffer<char>&, fmt::v8::basic_string_view<char>, fmt::v8::basic_format_args<fmt::v8::basic_format_context<std::__1::conditional<std::is_same<fmt::v8::type_identity<char>::type, char>::value, fmt::v8::appender, std::__1::back_insert_iterator<fmt::v8::detail::buffer<fmt::v8::type_identity<char>::type> > >::type, fmt::v8::type_identity<char>::type> >, fmt::v8::detail::locale_ref)::format_handler&) fmt/include/fmt/core.h:2546:23
  | #11 0x4b6311 in void fmt::v8::detail::parse_format_string<false, char, void fmt::v8::detail::vformat_to<char>(fmt::v8::detail::buffer<char>&, fmt::v8::basic_string_view<char>, fmt::v8::basic_format_args<fmt::v8::basic_format_context<std::__1::conditional<std::is_same<fmt::v8::type_identity<char>::type, char>::value, fmt::v8::appender, std::__1::back_insert_iterator<fmt::v8::detail::buffer<fmt::v8::type_identity<char>::type> > >::type, fmt::v8::type_identity<char>::type> >, fmt::v8::detail::locale_ref)::format_handler>(fmt::v8::basic_string_view<char>, void fmt::v8::detail::vformat_to<char>(fmt::v8::detail::buffer<char>&, fmt::v8::basic_string_view<char>, fmt::v8::basic_format_args<fmt::v8::basic_format_context<std::__1::conditional<std::is_same<fmt::v8::type_identity<char>::type, char>::value, fmt::v8::appender, std::__1::back_insert_iterator<fmt::v8::detail::buffer<fmt::v8::type_identity<char>::type> > >::type, fmt::v8::type_identity<char>::type> >, fmt::v8::detail::locale_ref)::format_handler&&) fmt/include/fmt/core.h:2571:21
  | #12 0x4b6311 in void fmt::v8::detail::vformat_to<char>(fmt::v8::detail::buffer<char>&, fmt::v8::basic_string_view<char>, fmt::v8::basic_format_args<fmt::v8::basic_format_context<std::__1::conditional<std::is_same<fmt::v8::type_identity<char>::type, char>::value, fmt::v8::appender, std::__1::back_insert_iterator<fmt::v8::detail::buffer<fmt::v8::type_identity<char>::type> > >::type, fmt::v8::type_identity<char>::type> >, fmt::v8::detail::locale_ref) fmt/include/fmt/format.h:2952:3
  | #13 0x4d18db in fmt::v8::appender fmt::v8::format_to<tm const&, 500ul, std::__1::allocator<char> >(fmt::v8::basic_memory_buffer<char, 500ul, std::__1::allocator<char> >&, fmt::v8::basic_format_string<char, fmt::v8::type_identity<tm const&>::type>, tm const&) fmt/include/fmt/format.h:3039:3
  | #14 0x4b5155 in void invoke_fmt<tm, long>(unsigned char const*, unsigned long) fmt/test/fuzzing/one-arg.cc:34:5
  | #15 0x4b3765 in LLVMFuzzerTestOneInput fmt/test/fuzzing/one-arg.cc:88:5
  | #16 0x43fd23 in fuzzer::Fuzzer::ExecuteCallback(unsigned char const*, unsigned long) cxa_noexception.cpp:0
  | #17 0x42b5d2 in fuzzer::RunOneTest(fuzzer::Fuzzer*, char const*, unsigned long) /src/llvm-project/compiler-rt/lib/fuzzer/FuzzerDriver.cpp:324:6
  | #18 0x431085 in fuzzer::FuzzerDriver(int*, char***, int (*)(unsigned char const*, unsigned long)) cxa_noexception.cpp:0
  | #19 0x459d82 in main /src/llvm-project/compiler-rt/lib/fuzzer/FuzzerMain.cpp:20:10
  | #20 0x7fa740d900b2 in __libc_start_main /build/glibc-eX1tMB/glibc-2.31/csu/libc-start.c:308:16
  | #21 0x40886d in _start
```

**Comment by phprus** [comment]
Input data is unknown?

I will prepare a patch today...

**Comment by phprus** [comment]
Cast to ``int64_t`` should be fix this UB (UB only for years over ``+-10^9``)
https://github.com/fmtlib/fmt/pull/2551

**Comment by vitaut** [comment]
Sanitizer caught one more UB due to overflow:

```c++
#include <fmt/chrono.h>

int main() {
  char tz[] = "PDT";
  auto t = std::tm{.tm_sec = 23,
                   .tm_min = 22,
                   .tm_hour = 22,
                   .tm_mday = 15,
                   .tm_mon = 9,
                   .tm_year = 2147483638,
                   .tm_wday = 6,
                   .tm_yday = 287,
                   .tm_isdst = 1,
                   .tm_gmtoff = -25200,
                   .tm_zone = tz};

  fmt::print("{:%F}", t);
}
```
